### PR TITLE
Add data driven SELF TEST code for signatures and key agreement

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Will need to be rebased after #11010 has been merged..

**Comment by slontis** [comment]
ping
Now that all algorithms are available in the FIPS module - the self tests have been updated.

**Comment by slontis** [comment]
ping

File: providers/fips/self_test_data.inc [hunk 1]
```diff
@@ -422,3 +496,687 @@ static const ST_KAT_DRBG st_kat_drbg_tests[] =
         ITM(drbg_hmac_sha1_pr_expected)
     }
 };
+
+/* KEY EXCHANGE TEST DATA */
+
+#ifndef OPENSSL_NO_DH
+/* DH KAT */
+static const unsigned char dh_p_hex[] = {
```
Comment (t8m) [inline providers/fips/self_test_data.inc]: Why _hex suffix? This applies to other data and algos below as well.

File: providers/fips/self_test_data.inc [hunk 2]
```diff
@@ -422,3 +496,687 @@ static const ST_KAT_DRBG st_kat_drbg_tests[] =
...
+    0x89, 0x78, 0xac, 0xa5, 0xb9, 0x44, 0x72, 0xf3,
+    0x39, 0xca, 0xeb, 0x99, 0x31, 0xb4, 0x2b, 0xe3,
+    0x01, 0x26, 0x8b, 0xc9, 0x97, 0x89, 0xc9, 0xb2,
+    0x55, 0x71, 0xc3, 0xc0, 0xe4, 0xcb, 0x3f, 0x00,
+    0x7f, 0x1a, 0x51, 0x1c, 0xbb, 0x53, 0xc8, 0x51,
+    0x9c, 0xdd, 0x13, 0x02, 0xab, 0xca, 0x6c, 0x0f,
+    0x34, 0xf9, 0x67, 0x39, 0xf1, 0x7f, 0xf4, 0x8b
+};
+
+static const unsigned char dh_secret_exptd[] = {
```
Comment (t8m) [inline providers/fips/self_test_data.inc]: _exptd looks really ugly - _expected is not that much longer or use _exp or even nothing.

File: providers/fips/self_test_data.inc [hunk 3]
```diff
@@ -422,3 +496,687 @@ static const ST_KAT_DRBG st_kat_drbg_tests[] =
...
+    ST_KAT_PARAM_END()
+};
+#endif /* OPENSSL_NO_DH */
+
+
+#ifndef OPENSSL_NO_EC
+static const char ecdh_curve_name[] = "prime256v1";
+static const unsigned char ecdh_privd_hex[] = {
+    0x33, 0xd0, 0x43, 0x83, 0xa9, 0x89, 0x56, 0x03,
+    0xd2, 0xd7,  0xfe, 0x6b, 0x01, 0x6f, 0xe4, 0x59,
```
Comment (t8m) [inline providers/fips/self_test_data.inc]: Extra space here.

File: providers/fips/self_test_data.inc [hunk 4]
```diff
@@ -422,3 +496,687 @@ static const ST_KAT_DRBG st_kat_drbg_tests[] =
...
+#endif /* OPENSSL_NO_EC */
+#ifndef OPENSSL_NO_DSA
+    {
+        OSSL_SELF_TEST_DESC_SIGN_DSA,
+        "DSA",
+        "SHA-256",
+        dsa_key,
+        /*
+         * The DSA signature changes each time due to it using a random k.
+         * So the there is no expected KAT for this..
```
Comment (t8m) [inline providers/fips/self_test_data.inc]: Two dots?

File: providers/fips/self_test_kats.c [hunk 5]
```diff
@@ -142,39 +144,78 @@ static int self_test_cipher(const ST_KAT_CIPHER *t, OSSL_SELF_TEST *st,
...
+    int ret = 0;
+    const ST_KAT_PARAM *p;
+
+    if (params == NULL)
+        return 1;
+    for (p = params; p->data != NULL; ++p)
+    {
+        switch (p->type) {
+        case OSSL_PARAM_UNSIGNED_INTEGER: {
+            BIGNUM *bn = BN_CTX_get(ctx); /* grab a bn from the pool */
```
Comment (t8m) [inline providers/fips/self_test_kats.c]: This comment feels a little bit superfluous.

Comment (slontis) [reply providers/fips/self_test_kats.c]: removed

File: providers/fips/self_test_data.inc (hunk 2 above)
Comment (slontis) [reply providers/fips/self_test_data.inc]: changed to _expected (The name indicates that it is an output parameter)

File: providers/fips/self_test_data.inc (hunk 1 above)
Comment (slontis) [reply providers/fips/self_test_data.inc]: removed

**Comment by slontis** [comment]
Had to rebase because OSSLPARAM_BLD API 's are now public.

**Comment by t8m** [comment]
The Travis failure is relevant.

**Comment by slontis** [comment]
Was missing a single = NULL - Should be good now.

**Comment by t8m** [comment]
Yep, now the Travis is just the usual timeout.

**Comment by slontis** [comment]
travis error was a timeout..

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**